<?xml version="1.0"?>
<launch>
  <arg name="use_sim_time" default="true" />
  <arg name="gui" default="true" />
  <arg name="headless" default="false" />

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="debug" value="0" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="use_sim_time" value="$(arg use_sim_time)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="paused" value="false"/>
  </include>

  <!-- Ideal formation position -->
  <rosparam param="formation_positions">[[0, 2, 0.5], [0, 1, 0.5], [0, 0, 0.5], [1, 2, 0.5], [1, 0, 0.5]]</rosparam>

  <!-- CBF Communication maintenance (+1) or obstacle avoidance (-1) -->
  <param name="safe_distance" value="2"/>
  <param name="cbf_direction" value="1"/>

  <!-- Nexus 1  -->
  <group ns="nexus1">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus1 -param robot_description -x -2 -y 4 -z 0.5" />
      <!-- Auto node -->
    <node name="cbf_formation_controller" pkg="first_task_formation_cbf" type="formation_keeping.py" output="screen" >
      <param name="robot_name" value="nexus1"/>
      <rosparam param="neighbours">[2]</rosparam>
    </node>
  </group>

  <!-- Nexus 2  -->
  <group ns="nexus2">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus2 -param robot_description -x -2 -y 2 -z 0.5" />
      <!-- Auto node -->
    <node name="cbf_formation_controller" pkg="first_task_formation_cbf" type="formation_keeping.py" output="screen" >
      <param name="robot_name" value="nexus2"/>
      <rosparam param="neighbours">[1, 3, 4, 5]</rosparam>
    </node>
  </group>

  <!-- Nexus 3  -->
  <group ns="nexus3">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus3 -param robot_description -x -3 -y -1 -z 0.5" />
      <!-- Auto node -->
    <node name="cbf_formation_controller" pkg="first_task_formation_cbf" type="formation_keeping.py" output="screen" >
      <param name="robot_name" value="nexus3"/>
      <rosparam param="neighbours">[2]</rosparam>
    </node>
  </group>

  <!-- Nexus 4  -->
  <group ns="nexus4">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus4 -param robot_description -x 3 -y 3 -z 0.5" />
      <!-- Auto node -->
    <node name="cbf_formation_controller" pkg="first_task_formation_cbf" type="formation_keeping.py" output="screen" >
      <param name="robot_name" value="nexus4"/>
      <rosparam param="neighbours">[2]</rosparam>
    </node>
  </group>

  <!-- Nexus 5  -->
  <group ns="nexus5">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus5 -param robot_description -x 5 -y -2 -z 0.5" />
      <!-- Auto node -->
    <node name="cbf_formation_controller" pkg="first_task_formation_cbf" type="formation_keeping.py" output="screen" >
      <param name="robot_name" value="nexus5"/>
      <rosparam param="neighbours">[2]</rosparam>
    </node>
  </group>

  <!-- Motion capture system simulation -->
  <include file="$(find mocap_simulator)/launch/qualisys_simulator.launch" />

</launch>